packages/-ember-data/tests/integration/record-data/record-data-test.ts

```diff
@@ -323,4 +367,294 @@ module('integration/record-data - Custom RecordData Implementations', function (
+        this._storeWrapper = storeWrapper;
+      }
+
+      getBelongsTo(key: string) {
+        assert.equal(key, 'landlord', 'Passed correct key to getBelongsTo');
+        return belongsToReturnValue;
+      }
+
+      setDirtyBelongsTo(key: string, recordData: this | null) {
+        assert.equal(key, 'landlord', 'Passed correct key to setBelongsTo');
+        assert.equal(recordData.id, '2', 'Passed correct RD to setBelongsTo');
+      }
+    }
+
+    let TestStore = Store.extend({
+      createRecordDataFor(modelName, id, clientId, storeWrapper) {
+        if (modelName === 'house') {
+          return new RelationshipRecordData(storeWrapper);
+        } else {
+          return this._super(modelName, id, clientId, storeWrapper);
+        }
+      }
+    });
+
+    owner.register('service:store', TestStore);
+
+    store = owner.lookup('service:store');
+
+    store.push({
+      data: [davidHash, runspiredHash]
+    });
+
+    store.push({
+      data: [houseHash]
+    });
+
+    let house = store.peekRecord('house', '1');
+    let runspired = store.peekRecord('person', '2');
+    assert.equal(house.get('landlord.name'), 'David', 'belongsTo get correctly looked up');
+
+    house.set('landlord', runspired);
+    assert.equal(house.get('landlord.name'), 'David', 'belongsTo does not change if RD did not notify');
+  });
+
+  test("Record Data custom belongsTo", async function (assert) {
+    assert.expect(4);
+    let { owner } = this;
+
+    let belongsToReturnValue = { data: { id: '1', type: 'person' } };
+
+    class RelationshipRecordData extends TestRecordData {
+      constructor(storeWrapper) {
+        super();
+        this._storeWrapper = storeWrapper;
+      }
+
+      getBelongsTo(key: string) {
+        assert.equal(key, 'landlord', 'Passed correct key to getBelongsTo');
+        return belongsToReturnValue;
+      }
+
+      setDirtyBelongsTo(key: string, recordData: this | null) {
+        belongsToReturnValue = { data: { id: '3', type: 'person' } };
+        this._storeWrapper.notifyBelongsToChange('house', '1', null, 'landlord');
+      }
+    }
+
+    let TestStore = Store.extend({
+      createRecordDataFor(modelName, id, clientId, storeWrapper) {
+        if (modelName === 'house') {
+          return new RelationshipRecordData(storeWrapper);
+        } else {
+          return this._super(modelName, id, clientId, storeWrapper);
+        }
+      }
+    });
+
+    owner.register('service:store', TestStore);
+
+    store = owner.lookup('service:store');
+
+    store.push({
+      data: [davidHash, runspiredHash, igorHash]
+    });
+
+    store.push({
+      data: [houseHash]
+    });
+
+    let house = store.peekRecord('house', '1');
+    assert.equal(house.get('landlord.name'), 'David', 'belongsTo get correctly looked up');
+
+    let runspired = store.peekRecord('person', '2');
+    house.set('landlord', runspired);
+
+    assert.equal(house.get('landlord.name'), 'Igor', 'RecordData sets the custom belongsTo value');
```

Comment:
also note the intentional misdirection here

Reply:
fixed at call site